tsgo panics when a get/set accessor with an illegal async modifier contains for-await-of

## Stack trace

```
panic: Enclosing function should never be an async function. [recovered, repanicked]

goroutine 307 [running]:
sync.(*WaitGroup).Go.func1.1()
        sync/waitgroup.go:251 +0x48
panic({0x1057e8060?, 0x105a88c50?})
        runtime/panic.go:860 +0x12c
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkGrammarForInOrForOfStatement(0x16e71c349908, 0x16e71bb5e000)
        github.com/microsoft/typescript-go/internal/checker/grammarchecks.go:1250 +0x6a4
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkForOfStatement(0x16e71c349908, 0x16e71bb5e000)
        github.com/microsoft/typescript-go/internal/checker/checker.go:4006 +0x44
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElementWorker(0x16e71c349908, 0x16e71bb5e000)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2320 +0x2dc
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElement(0x16e71c349908, 0x10585cbc0?)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2225 +0x58
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElements(0x16e71c349908, {0x16e71bb380e0, 0x1, 0x1048e6448?})
        github.com/microsoft/typescript-go/internal/checker/checker.go:2215 +0x34
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkBlock(0x16e71c349908, 0x16e71bb34120)
        github.com/microsoft/typescript-go/internal/checker/checker.go:3764 +0x12c
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElementWorker(0x16e71c349908, 0x16e71bb34120)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2304 +0x354
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElement(0x16e71c349908, 0x16e71bb5a130?)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2225 +0x58
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkAccessorDeclaration(0x16e71c349908, 0x16e71bb5a130)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2956 +0x2a8
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElementWorker(0x16e71c349908, 0x16e71bb5a130)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2266 +0x144
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElement(0x16e71c349908, 0x1048df258?)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2225 +0x58
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElements(0x16e71c349908, {0x16e71bb380e8, 0x1, 0x1048de47c?})
        github.com/microsoft/typescript-go/internal/checker/checker.go:2215 +0x34
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkClassDeclaration(0x16e71c349908, 0x16e71bb3a128)
        github.com/microsoft/typescript-go/internal/checker/checker.go:4258 +0x13c
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElementWorker(0x16e71c349908, 0x16e71bb3a128)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2340 +0x334
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElement(0x16e71c349908, 0x16e71d6c3e60?)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2225 +0x58
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceElements(0x16e71c349908, {0x16e71bb380f0, 0x1, 0x10478b874?})
        github.com/microsoft/typescript-go/internal/checker/checker.go:2215 +0x34
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceFile(0x16e71c349908, {0x105a946e0?, 0x105c952a0?}, 0x16e71bb62008, 0x0)
        github.com/microsoft/typescript-go/internal/checker/checker.go:2183 +0x19c
github.com/microsoft/typescript-go/internal/checker.(*Checker).getDiagnostics(0x16e71c349908, {0x105a946e0?, 0x105c952a0?}, 0x16e71bb62008, 0x16e71c349b00)
        github.com/microsoft/typescript-go/internal/checker/checker.go:13847 +0x70
github.com/microsoft/typescript-go/internal/checker.(*Checker).GetDiagnostics(...)
        github.com/microsoft/typescript-go/internal/checker/checker.go:13837
github.com/microsoft/typescript-go/internal/compiler.(*Program).getBindAndCheckDiagnosticsWithChecker(0x16e71bbba788, {0x105a946e0, 0x105c952a0}, 0x16e71c349908, 0x16e71bb62008)
        github.com/microsoft/typescript-go/internal/compiler/program.go:1307 +0x90
github.com/microsoft/typescript-go/internal/compiler.(*Program).getSemanticDiagnosticsWithChecker(0x16e71bbba788, {0x105a946e0?, 0x105c952a0?}, 0x1047e6a70?, 0x16e71bb62008)
        github.com/microsoft/typescript-go/internal/compiler/program.go:1291 +0x2c
github.com/microsoft/typescript-go/internal/compiler.(*Program).collectCheckerDiagnosticsFromFiles.func1(0x105852a00?, 0x52, 0x16e71bb62008?)
        github.com/microsoft/typescript-go/internal/compiler/program.go:554 +0x44
github.com/microsoft/typescript-go/internal/compiler.(*checkerPool).forEachCheckerGroupDo.func1()
        github.com/microsoft/typescript-go/internal/compiler/checkerpool.go:159 +0x168
github.com/microsoft/typescript-go/internal/core.(*parallelWorkGroup).Queue.func1()
        github.com/microsoft/typescript-go/internal/core/workgroup.go:40 +0x24                     sync.(*WaitGroup).Go.func1()
        sync/waitgroup.go:258 +0x48
created by sync.(*WaitGroup).Go in goroutine 1                                                             sync/waitgroup.go:238 +0x70
```

## Steps to reproduce

```ts
class C {
  async get x() {
    for await (const y of []) {
      // ...
    }
  }
}
```

--- a/internal/checker/grammarchecks.go
+++ b/internal/checker/grammarchecks.go
@@ -1246,9 +1246,6 @@ func (c *Checker) checkGrammarForInOrForOfStatement(forInOrOfStatement *ast.ForI
 					containingFunc := ast.GetContainingFunction(forInOrOfStatement.AsNode())
 					if containingFunc != nil && containingFunc.Kind != ast.KindConstructor {
 						debug.Assert((ast.GetFunctionFlags(containingFunc)&ast.FunctionFlagsAsync) == 0, "Enclosing function should never be an async function.")
-						if hasAsyncModifier(containingFunc) {
-							panic("Enclosing function should never be an async function.")
-						}
 						relatedInfo := createDiagnosticForNode(containingFunc, diagnostics.Did_you_mean_to_mark_this_function_as_async)
 						diagnostic.AddRelatedInfo(relatedInfo)
 					}

--- a/testdata/baselines/reference/compiler/forAwaitInAsyncAccessor.errors.txt
+++ b/testdata/baselines/reference/compiler/forAwaitInAsyncAccessor.errors.txt
@@ -1,0 +1,20 @@
+a.ts(2,5): error TS1042: 'async' modifier cannot be used here.
+a.ts(2,15): error TS2378: A 'get' accessor must return a value.
+a.ts(3,13): error TS1103: 'for await' loops are only allowed within async functions and at the top levels of modules.
+
+
+==== a.ts (3 errors) ====
+    class C {
+        async get x() {
+        ~~~~~
+!!! error TS1042: 'async' modifier cannot be used here.
+                  ~
+!!! error TS2378: A 'get' accessor must return a value.
+            for await (const y of []) {
+                ~~~~~
+!!! error TS1103: 'for await' loops are only allowed within async functions and at the top levels of modules.
+!!! related TS1356 a.ts:2:15: Did you mean to mark this function as 'async'?
+            }
+        }
+    }
+    

--- a/testdata/baselines/reference/compiler/forAwaitInAsyncAccessor.js
+++ b/testdata/baselines/reference/compiler/forAwaitInAsyncAccessor.js
@@ -1,0 +1,19 @@
+//// [tests/cases/compiler/forAwaitInAsyncAccessor.ts] ////
+
+//// [a.ts]
+class C {
+    async get x() {
+        for await (const y of []) {
+        }
+    }
+}
+
+
+//// [a.js]
+"use strict";
+class C {
+    async get x() {
+        for await (const y of []) {
+        }
+    }
+}

--- a/testdata/baselines/reference/compiler/forAwaitInAsyncAccessor.symbols
+++ b/testdata/baselines/reference/compiler/forAwaitInAsyncAccessor.symbols
@@ -1,0 +1,15 @@
+//// [tests/cases/compiler/forAwaitInAsyncAccessor.ts] ////
+
+=== a.ts ===
+class C {
+>C : Symbol(C, Decl(a.ts, 0, 0))
+
+    async get x() {
+>x : Symbol(C.x, Decl(a.ts, 0, 9))
+
+        for await (const y of []) {
+>y : Symbol(y, Decl(a.ts, 2, 24))
+        }
+    }
+}
+

--- a/testdata/baselines/reference/compiler/forAwaitInAsyncAccessor.types
+++ b/testdata/baselines/reference/compiler/forAwaitInAsyncAccessor.types
@@ -1,0 +1,16 @@
+//// [tests/cases/compiler/forAwaitInAsyncAccessor.ts] ////
+
+=== a.ts ===
+class C {
+>C : C
+
+    async get x() {
+>x : void
+
+        for await (const y of []) {
+>y : never
+>[] : never[]
+        }
+    }
+}
+

--- a/testdata/tests/cases/compiler/forAwaitInAsyncAccessor.ts
+++ b/testdata/tests/cases/compiler/forAwaitInAsyncAccessor.ts
@@ -1,0 +1,7 @@
+// @filename: a.ts
+class C {
+    async get x() {
+        for await (const y of []) {
+        }
+    }
+}

